### ISSUE TITLE
migrate testing to AVA

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#1568]
Our current best-looking result from the #1335 survey is to use [AVA](https://github.com/avajs/ava) instead of tap/tape.

This ticket is about performing that migration.

Questions for me to answer first:

* does AVA play nicely with ESM (without `-r esm`)
  * we'll be adding an AVA config to add `-r esm` for all packages that are not yet set to `type: module`
* what will be our new best-practices for adding `console.log` to debug test problems
* write up instructions to run just one test
* write up instructions to run under the debugger

I'm hoping to find a way to convert one modules at a time, for smaller PRs, but I won't let that stop me.

I'll be drawing from @dtribble 's #1447 PR.

- [x] assert
- [x] install-ses
- [x] import-manager
- [x] sparse-ints
- [x] registrar
- [x] store
- [x] weak-store
- [x] acorn-eventual-send
- [x] bundle-source
- [x] import-bundle
- [x] eventual-send
- [x] promise-kit
- [x] transform-eventual-send
- [x] tame-metering
- [x] transform-metering
- [x] install-metering-and-ses
- [x] marshal
- [x] same-structure
- [x] captp
- [x] stat-logger
- [x] swing-store-lmdb
- [x] swing-store-simple
- [x] SwingSet
- [x] swingset-runner
- [x] ERTP
- [x] spawner
- [x] sharing-service
- [x] zoe
- [x] dapp-svelte-wallet/ui
- [x] dapp-svelte-wallet/api
- [x] dapp-svelte-wallet
- [x] cosmic-swingset
- [x] agoric-cli
- [x] deployment
- [x] notifier

**Comment by warner** [comment]
@michaelfig points out that we have non-test files with names like `testHelpers.js` that should not be treated as tests. So our `files` config should be `packages/*/test/**/test-*.js`, not `packages/*/test/**/test*.js`.

I'll look to see if we also have `test.js` anywhere (which would be appropriate for a small package that only has one test file). We'll either need to rename those to something longer, or augment `files` to a second glob of `packages/*/test/**/test.js`.

Either way, we need to update our coding-standards document to recommend the naming scheme.

**Comment by warner** [comment]
I'm going to proceed one package at a time, leaving the top-level `yarn test` mapped to `yarn workspaces test`, until all packages are converted. Then we can make a final change to the top-level `package.json` and enable full parallelism.

**Comment by warner** [comment]
I've checked a simple package (with no dependencies) with `"type": "module"` under AVA (removing the `require: ["esm"]` from the AVA config) and it seems to work. I also tried the same *with* the `require: ["esm"]` and it still works, which suggests that maybe `-r esm` is more tolerant of actually-declared-ESM files than I thought, and offers some hope for the heterogenous-packages conflict described above.

**Comment by warner** [comment]
AVA rejects `ava --config ../../ava-shared-config.js`: it's really insistent upon having the config file live next to the nearest `package.json`. That rules out a shared config, without resorting to symlinks or pre-test steps that create that local file (by copying it from a shared source). It's probably not a big deal, there aren't a lot of options to change anyways, and you can always make per-invocation changes by just adding them to the `yarn` script command line (e.g. `yarn test --verbose` or `yarn test -m '*testname to run*'`).

**Comment by warner** [comment]
All packages have been converted. The remaining step is make the top-level `yarn test` target use AVA in parallel across all packages, rather than doing a `yarn workspaces test` which runs one package at a time.

This might improve reporting: AVA would be in control of all tests. Success/failure/progress would be reported one test file at a time. AVA would not even be aware of what package contains each test file. And it could improve performance slightly, by increasing the opportunity for parallelism.

The complication is that we can't use the same AVA config for all our tests, specifically the `ava.require=[]` list. We need `-r esm` for everything (for now). Of our 109 `test-*.js` files, we need `@agoric/install-ses` for 83 of them, `@agoric/install-metering-and-ses` for 10, and neither for 16. @dtribble added `ava.require=['esm', '@agoric/install-ses']` for the 6 in ERTP with good results, and we could do that for most of the rest (adding just SES isn't likely to cause any problems).

However the 10 that actually exercise metering couldn't be included under that command. We'd have to have two separate `ava` invocations, one with the default (using the `package.json` config of `ava.require=['esm', '@agoric/install-ses']`), and a second with an alternate config file that used `ava.require=['esm', '@agoric/install-metering-and-ses']`. These would not run in parallel: we'd do all the non-metering tests in one pass, then all the metering tests in a second pass. We'd move all the metering-based tests into separate directories, probably `PACKAGENAME/test/*` and `PACKAGENAME/metering-test/*`, so the top-level config file could find the right ones.

We'd need to remove all the manual `import '@agoric/install-ses'` / `import '@agoric/install-metering-and-ses'` lines from our test files, which means they can only be run under AVA (which is the case anyways, because they `import test from 'ava'`, which balks otherwise). but it would also change our process for running a single test file: instead of just `yarn test test/test-foo.js`, if the test is a metering test, you'd need to run something like `yarn test-metering metering-test/test-bar.js`. Running `yarn test metering-test/test-bar.js` would result in an error becauase the metering code would not get installed.

**Comment by warner** [comment]
We no longer use `install-metering-and-ses`, so I think that blocker is gone.

**Comment by warner** [comment]
closing this in favor of the smaller #4493
